plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java

```diff
@@ -140,17 +144,45 @@ public ConnectorTableMetadata getTableMetadata()
         return connectorTableMetadata;
     }
 
-    private Optional<RowType> getPartitionColumnType(List<PartitionField> fields, Schema schema)
+    private static List<PartitionField> getAllPartitionFields(Table icebergTable)
+    {
+        Set<Integer> existingColumnsIds = icebergTable.schema()
+                .columns().stream()
+                .map(Types.NestedField::fieldId)
+                .collect(toUnmodifiableSet());
+
+        List<PartitionField> visiblePartitionFields = icebergTable.specs()
+                .values().stream()
+                .flatMap(partitionSpec -> partitionSpec.fields().stream())
+                // skip columns that were dropped
+                .filter(partitionField -> existingColumnsIds.contains(partitionField.sourceId()))
+                .collect(toImmutableList());
+
+        Set<Integer> alreadyExistingFieldIds = new HashSet<>();
+        List<PartitionField> result = new ArrayList<>();
+        for (PartitionField partitionField : visiblePartitionFields) {
+            if (!alreadyExistingFieldIds.contains(partitionField.fieldId())) {
+                alreadyExistingFieldIds.add(partitionField.fieldId());
+                result.add(partitionField);
+            }
+        }
+        return result;
+    }
+
+    private Optional<IcebergPartitionColumn> getPartitionColumnType(List<PartitionField> fields, Schema schema)
     {
         List<RowType.Field> partitionFields = fields.stream()
                 .map(field -> RowType.field(
                         field.name(),
                         toTrinoType(field.transform().getResultType(schema.findType(field.sourceId())), typeManager)))
                 .collect(toImmutableList());
+        List<Integer> ids = fields.stream()
+                .map(PartitionField::fieldId)
+                .collect(toImmutableList());
         if (partitionFields.isEmpty()) {
```

Comment:
unrelated to the current commit:
```
if (partitionFields.isEmpty()) {
            return Optional.empty();
}
```

this can be replaced with a check 

```
if (fields.isEmpty()) {
            return Optional.empty();
}
```
at the beginning of the method.

Reply:
I will add it as a separate commit